src/browser/screen.js

```diff
@@ -487,20 +507,52 @@ function ScreenAdapter(screen_container, bus)
 
     this.update_buffer = function(min, max)
     {
-        if(max < min)
+        if(max < min && !layers_has_changed)
         {
             return;
         }
 
-        var min_y = min / graphical_mode_width | 0;
-        var max_y = max / graphical_mode_width | 0;
+        layers_has_changed = false;
 
-        graphic_context.putImageData(
-            graphic_image_data,
-            0, 0,
-            0, min_y,
-            graphical_mode_width, max_y - min_y + 1
-        );
+        this.clear_screen();
+        if(DEBUG_SCREEN_LAYERS)
+        {
+            // Draw the entire buffer. Useful for debugging
+            // panning / page flipping / screen splitting code for both
+            // v86 developers and os developers
+            graphic_context.putImageData(
+                graphic_image_data,
+                0, 0
+            );
+
+            // For each visible layer that would've been drawn, draw a
+            // rectangle to visualise the layer instead.
+            graphic_context.strokeStyle = "#0F0";
+            graphic_context.lineWidth = 4;
+            layers.forEach((layer) =>
+            {
+                graphic_context.strokeRect(
+                    layer.buffer_x,
+                    layer.buffer_y,
+                    layer.buffer_width,
+                    layer.buffer_height
+                );
+            });
+            graphic_context.lineWidth = 1;
+            return;
+        }
+        layers.forEach((layer) =>
+        {
+            graphic_context.putImageData(
```

Comment:
> Can we restore the old behaviour at least for svga modes?

Sure. Since there are only a couple of layers at most, adding some logic to restrict the drawing of each layer shouldn't leave much impact.

> If possible, we should move the layer handling to vga.js, since this file is for platform-independent rendering (but since the layer logic is quite simple we could leave it here).

I think I get what you mean. I'll first take account of the `min` and `max` in `screen.js` while keeping `screen.js` device-agnostic and platform-agnostic. After that, I think I'll try implementing the layering in `vga.js` to compare the performances.

Reply:
>Sure. Since there are only a couple of layers at most, adding some logic to restrict the drawing of each layer shouldn't leave much impact.

The logic required to update the right pixels, given that the layer sizes can change over time, is a bit more complicated than I thought. Not only do we need to update the pixels between `min` and `max` in the large image buffer, but also the pixels on the canvas where the layer boundaries have moved or changed.

For svga, I'm thinking that we can use the layers to implement the `min` and `max`. I.e., instead of sending the `min` and `max` values, we can send a layer object representing the pixels that needs updating for svga.